File: apps/cli/src/agent-command.ts
import { constants } from "node:fs";
import { access } from "node:fs/promises";
import { homedir } from "node:os";
import { delimiter, join } from "node:path";

const COMMON_BIN_DIRS = [
  "/opt/homebrew/bin",
  "/usr/local/bin",
  "/usr/bin",
  join(homedir(), ".local", "bin"),
  join(homedir(), ".nvm", "current", "bin"),
];

const APP_BINARIES: Record<"codex" | "claude", string[]> = {
  codex: [
    "/Applications/ChatGPT.app/Contents/Resources/codex",
    "/Applications/Codex.app/Contents/Resources/codex",
  ],
  claude: [
    "/Applications/Claude.app/Contents/Resources/claude",
  ],
};

async function executable(path: string): Promise<boolean> {
  try {
    await access(path, constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

export async function resolveAgentCommand(command: "codex" | "claude"): Promise<string | null> {
  const override = process.env[command === "codex" ? "ZIMLO_CODEX_BIN" : "ZIMLO_CLAUDE_BIN"]?.trim();
  if (override) return await executable(override) ? override : null;

  const pathDirs = (process.env.PATH ?? "").split(delimiter).filter(Boolean);
  const candidates = [
    ...APP_BINARIES[command],
    ...pathDirs.map((directory) => join(directory, command)),
    ...COMMON_BIN_DIRS.map((directory) => join(directory, command)),
  ];
  for (const candidate of [...new Set(candidates)]) {
    if (await executable(candidate)) return candidate;
  }
  return null;
}

export async function requireAgentCommand(command: "codex" | "claude"): Promise<string> {
  const resolved = await resolveAgentCommand(command);
  if (resolved) return resolved;

  const label = command === "codex" ? "Codex" : "Claude Code";
  throw new Error(`未找到 ${label} Runtime。请确认应用已安装，或在 Zimlo 设置中检查 Runtime 接入。`);
}

export async function detectInstalledProviders(): Promise<Array<"codex" | "claude">> {
  const [codexCommand, claudeCommand] = await Promise.all([
    resolveAgentCommand("codex"),
    resolveAgentCommand("claude"),
  ]);
  return [
    ...(codexCommand ? ["codex" as const] : []),
    ...(claudeCommand ? ["claude" as const] : []),
  ];
}

File: runtime/crates/zimlo-bridge/src/agent_command.rs
use std::{collections::HashSet, path::PathBuf};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum AgentProvider {
    Codex,
    Claude,
}

impl AgentProvider {
    fn command(self) -> &'static str {
        match self {
            Self::Codex => "codex",
            Self::Claude => "claude",
        }
    }

    fn override_name(self) -> &'static str {
        match self {
            Self::Codex => "ZIMLO_CODEX_BIN",
            Self::Claude => "ZIMLO_CLAUDE_BIN",
        }
    }

    fn application_binaries(self) -> &'static [&'static str] {
        match self {
            Self::Codex => &[
                "/Applications/ChatGPT.app/Contents/Resources/codex",
                "/Applications/Codex.app/Contents/Resources/codex",
            ],
            Self::Claude => &["/Applications/Claude.app/Contents/Resources/claude"],
        }
    }
}

pub(crate) async fn resolve(provider: AgentProvider) -> Option<PathBuf> {
    if let Some(path) = std::env::var_os(provider.override_name())
        .map(PathBuf::from)
        .filter(|path| !path.as_os_str().is_empty())
    {
        return executable(&path).await.then_some(path);
    }

    let mut candidates = provider
        .application_binaries()
        .iter()
        .map(PathBuf::from)
        .collect::<Vec<_>>();
    if let Some(path) = std::env::var_os("PATH") {
        candidates.extend(std::env::split_paths(&path).map(|path| path.join(provider.command())));
    }
    candidates.extend(
        ["/opt/homebrew/bin", "/usr/local/bin", "/usr/bin"]
            .map(PathBuf::from)
            .map(|path| path.join(provider.command())),
    );
    if let Some(home) = std::env::var_os("HOME").map(PathBuf::from) {
        candidates.push(home.join(".local/bin").join(provider.command()));
        candidates.push(home.join(".nvm/current/bin").join(provider.command()));
    }

    let mut seen = HashSet::new();
    for candidate in candidates {
        if seen.insert(candidate.clone()) && executable(&candidate).await {
            return Some(candidate);
        }
    }
    None
}

async fn executable(path: &PathBuf) -> bool {
    let Ok(metadata) = tokio::fs::metadata(path).await else {
        return false;
    };
    if !metadata.is_file() {
        return false;
    }
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt as _;
        metadata.permissions().mode() & 0o111 != 0
    }
    #[cfg(not(unix))]
    {
        true
    }
}
